### PR TITLE
*: manual pause a task which has been pause by error

### LINKED ISSUE #1729: Task auto resume after manually pause an error task.
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do? If possible, provide a recipe for reproducing the error.
upstream report an error after start/resume task immediately. (e.g. failed to create table)
task auto-resume
fail to execute `operate-schema` because task not pause
2. What did you expect to see?
manually pause task
exec `operate-schema set`
3. What did you see instead?
pause task fails because error immediately occur after auto resume, we can't pause a paused task.
operate-schema fails because task auto-resume since it's paused by error
4. Versions of the cluster

    - DM version (run `dmctl -V` or `dm-worker -V` or `dm-master -V`):

        ```
        master
        ```

### PR DESCRIPTION
**Pull request by GMHDBJD**
### What problem does this PR solve? 
close #1729 

### What is changed and how it works?
TaskChecker will not resume a task that is manually paused(isCancel: true)
When users manual pauses a task which has been pause by error(task stage is paused while isCancel is false), mark the isCancel to true.

### Check List 

Tests 

 - Unit test
 - Integration test
 - Manual test (add detailed scripts or steps below)
 - No code

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to update the `dm/dm-ansible`
 - Need to be included in the release note

### CODE REVIEW

File: dm/worker/subtask.go [hunk 1]
```diff
@@ -459,6 +475,10 @@ func (st *SubTask) Close() {
 
 // Pause pauses the running sub task.
```
Comment (Ehco1996) [inline dm/worker/subtask.go:476]: maybe update the comment

File: syncer/streamer_controller.go [hunk 2]
```diff
@@ -253,6 +256,12 @@ func (c *StreamerController) GetEvent(tctx *tcontext.Context) (event *replicatio
 	failpoint.Inject("GetEventError", func() {
 		err = errors.New("go-mysql returned an error")
 	})
+	failpoint.Inject("GetEventErrorOnce", func() {
+		if !mockGetEventErr {
```
Comment (Ehco1996) [inline syncer/streamer_controller.go:260]: why we need check `mockGetEventErr` ?

Comment (GMHDBJD) [reply syncer/streamer_controller.go:260]: Otherwise, the same error occurs when we manually resume-task.

Comment (Ehco1996) [reply syncer/streamer_controller.go:260]: oh i know, you don't want to remove failpoint and restart worker

File: dm/worker/subtask.go [hunk 3]
```diff
@@ -435,6 +435,22 @@ func (st *SubTask) setResult(result *pb.ProcessResult) {
 	st.result = result
 }
 
+// markResultCanceled mark result as canceled if stage is Paused.
+// This func is used to pause a task which has been paused by error,
```
Comment (lance6716) [inline dm/worker/subtask.go:439]: we should also support "pause a task which has no error, just the user want to pause it" 🤔

File: dm/worker/subtask.go [hunk 4]
```diff
@@ -457,8 +473,12 @@ func (st *SubTask) Close() {
 	updateTaskState(st.cfg.Name, st.cfg.SourceID, pb.Stage_Stopped)
 }
 
-// Pause pauses the running sub task.
+// Pause pauses a running sub task or a sub task paused by error.
 func (st *SubTask) Pause() error {
+	if st.markResultCanceled() {
```
Comment (lance6716) [inline dm/worker/subtask.go:478]: if `markResultCanceled` is called before chaning stage, user need to pause twice to pin the paused status. I think his expectation is pausing once

File: dm/worker/subtask.go (hunk 3 above)
Comment (GMHDBJD) [reply dm/worker/subtask.go:439]: I don't understand what you mean. We currently support pause-task manually for a normal task.

File: syncer/streamer_controller.go (hunk 2 above)
Comment (lichunzhu) [reply syncer/streamer_controller.go:260]: Can we do this by `"github.com/pingcap/dm/syncer/GetEventError=1*return"`?
Reference: https://github.com/pingcap/failpoint#key-concepts
Example: [dumpling failpoint](https://github.com/pingcap/dumpling/blob/843aff7/tests/basic/run.sh#L71)

File: tests/drop_column_with_index/run.sh [hunk 5]
```diff
@@ -33,9 +36,30 @@ function run() {
...
+		"pause-task test" \
+		"\"result\": true" 1 \
+		"get event returned an error" 1
+	run_dm_ctl_with_retry $WORK_DIR "127.0.0.1:$MASTER_PORT" \
+		"query-status test" \
+		"get event returned an error" 1 \
+		"\"stage\": \"Paused\"" 1 \
+		"\"isCanceled\": true" 1
+
+	sleep 5
```
Comment (lichunzhu) [inline tests/drop_column_with_index/run.sh:56]: why sleep 5 here?

File: dm/worker/subtask.go (hunk 4 above)
Comment (lichunzhu) [reply dm/worker/subtask.go:478]: +1

File: tests/drop_column_with_index/run.sh (hunk 5 above)
Comment (GMHDBJD) [reply tests/drop_column_with_index/run.sh:56]: auto resume interval 5 seconds

File: dm/worker/subtask.go (hunk 4 above)
Comment (GMHDBJD) [reply dm/worker/subtask.go:478]: If user pause a running task, DM will [callCurrentCancel](https://github.com/pingcap/dm/blob/068c30eb344c2556f13b1dbf7a6e55feed7ca1f3/dm/worker/subtask.go#L466) and [fetchResult](https://github.com/pingcap/dm/blob/068c30eb344c2556f13b1dbf7a6e55feed7ca1f3/dm/worker/subtask.go#L252) will mark the task paused and set isCanceled to true.
If task paused by error, isCanceled is false, so simply replace the value, and [task checker](https://github.com/pingcap/dm/blob/068c30eb344c2556f13b1dbf7a6e55feed7ca1f3/dm/worker/task_checker.go#L260) will not resume it.

Comment (lance6716) [reply dm/worker/subtask.go:478]: so the discription of https://github.com/pingcap/dm/issues/1729 is wrong? task will keep paused when user manually paused it, but just before manually pause, some error would occur and get a paused status, so we wrongly take it as manually pause didn't take effect?

Comment (GMHDBJD) [reply dm/worker/subtask.go:478]: Yes！

**Review by glorv — APPROVED** [review]

File: tests/drop_column_with_index/run.sh (hunk 5 above)
Comment (lance6716) [reply tests/drop_column_with_index/run.sh:56]: we may reduce check interval and sleep twice as it

Comment (GMHDBJD) [reply tests/drop_column_with_index/run.sh:56]: If check-interval is too small, the task may auto resume before manually pause, and test will be unstable.

**Review by lance6716 — APPROVED** [review]

**Pull request merged**